File: src/steering_path_lib/src/steering_path.cpp
#include "steering_path_lib/steering_path.h"

#include "steering_path_lib/path_integrator.h"
#include "steering_functions/all_in_one.hpp"
namespace steering
{

    SteeringPath::SteeringPath(PathType path_type,
                               double   kappa_max,
                               double   sigma_max,
                               float    discretization)
        : path_type_(path_type)
        , discretization_(discretization)
        , kappa_max_(kappa_max)
        , sigma_max_(sigma_max)
    {
        if (kappa_max <= 0 || sigma_max < 0 || discretization <= 0)
        {
            throw std::invalid_argument("Invalid parameters in SteeringPath constructor");
        }

        base_state_space_ = createPlanner();
        if (!base_state_space_)
        {
            throw std::runtime_error("Failed to create path planner");
        }
    }

    std::unique_ptr<BaseStateSpace> SteeringPath::createPlanner() const
    {
        switch (path_type_)
        {
        case PathType::CC_DUBINS:
            return std::make_unique<CC_Dubins_State_Space>(
                kappa_max_, sigma_max_, discretization_, true);
        case PathType::CC00_DUBINS:
            return std::make_unique<CC00_Dubins_State_Space>(
                kappa_max_, sigma_max_, discretization_, true);
        case PathType::CC0PM_DUBINS:
            return std::make_unique<CC0pm_Dubins_State_Space>(
                kappa_max_, sigma_max_, discretization_, true);
        case PathType::CCPM0_DUBINS:
            return std::make_unique<CCpm0_Dubins_State_Space>(
                kappa_max_, sigma_max_, discretization_, true);
        case PathType::CCPMPM_DUBINS:
            return std::make_unique<CCpmpm_Dubins_State_Space>(
                kappa_max_, sigma_max_, discretization_, true);
        case PathType::DUBINS:
            return std::make_unique<Dubins_State_Space>(kappa_max_, discretization_, true);
        case PathType::CC00_RS:
            return std::make_unique<CC00_Reeds_Shepp_State_Space>(
                kappa_max_, sigma_max_, discretization_);
        case PathType::HC_RS:
            return std::make_unique<HC_Reeds_Shepp_State_Space>(
                kappa_max_, sigma_max_, discretization_);
        case PathType::HC00_RS:
            return std::make_unique<HC00_Reeds_Shepp_State_Space>(
                kappa_max_, sigma_max_, discretization_);
        case PathType::HC0PM_RS:
            return std::make_unique<HC0pm_Reeds_Shepp_State_Space>(
                kappa_max_, sigma_max_, discretization_);
        case PathType::HCPM0_RS:
            return std::make_unique<HCpm0_Reeds_Shepp_State_Space>(
                kappa_max_, sigma_max_, discretization_);
        case PathType::HCPMPM_RS:
            return std::make_unique<HCpmpm_Reeds_Shepp_State_Space>(
                kappa_max_, sigma_max_, discretization_);
        case PathType::RS:
            return std::make_unique<Reeds_Shepp_State_Space>(kappa_max_, discretization_);
        default:
            return std::make_unique<Reeds_Shepp_State_Space>(kappa_max_, discretization_);
        }
    }

    std::vector<std::vector<Control>> SteeringPath::computeControlSequences(
        const State& state_start, const State& state_goal, bool use_shortest_path) const
    {
        if (!base_state_space_)
        {
            throw std::runtime_error("State space not initialized");
        }

        if (use_shortest_path)
        {
            std::vector<Control> controls;
            // For shortest path, compute single optimal trajectory
            base_state_space_->get_path(state_start, state_goal, controls);
            return {controls};
        }

        // Compute all possible control sequences between states
        return base_state_space_->get_all_controls(state_start, state_goal);
    }

    std::vector<std::vector<State>> SteeringPath::computeAllPaths(const State& state_start,
                                                                  const State& state_goal,
                                                                  bool use_shortest_path) const
    {
        if (!base_state_space_)
        {
            throw std::runtime_error("State space not initialized");
        }

        // Get all possible control sequences
        auto control_sequences =
            computeControlSequences(state_start, state_goal, use_shortest_path);

        // Generate discretized paths for each control sequence
        std::vector<std::vector<State>> all_paths;
        all_paths.reserve(control_sequences.size());

        for (const auto& controls : control_sequences)
        {
            all_paths.push_back(
                PathIntegrator::generateDiscretizedPath(state_start, controls, discretization_));
        }

        return all_paths;
    }

} // namespace steering

File: python_rewrite/steering_functions/steering_path.py
"""High-level SteeringPath wrapper and PathType enum (Python port)."""

from enum import Enum, auto
from steering_functions.state import State, Control
from steering_functions.base_state_space import BaseStateSpace


class PathType(Enum):
    """Supported path planning algorithm types."""
    NONE = 0
    CC_DUBINS = auto()
    CC00_DUBINS = auto()
    CC0PM_DUBINS = auto()
    CCPM0_DUBINS = auto()
    CCPMPM_DUBINS = auto()
    DUBINS = auto()
    CC00_RS = auto()
    HC_RS = auto()
    HC00_RS = auto()
    HC0PM_RS = auto()
    HCPM0_RS = auto()
    HCPMPM_RS = auto()
    RS = auto()


def _create_planner(path_type, kappa_max, sigma_max, discretization):
    """Factory: create the appropriate BaseStateSpace subclass."""

    if path_type == PathType.DUBINS:
        from steering_functions.dubins_state_space import DubinsStateSpace
        return DubinsStateSpace(kappa_max, discretization, True)

    if path_type == PathType.RS:
        from steering_functions.reeds_shepp_state_space import ReedsSheppStateSpace
        return ReedsSheppStateSpace(kappa_max, discretization)

    if path_type == PathType.CC_DUBINS:
        from steering_functions.cc_dubins_state_space import CC_Dubins_State_Space
        return CC_Dubins_State_Space(kappa_max, sigma_max, discretization, True)

    if path_type == PathType.CC00_DUBINS:
        from steering_functions.cc_dubins_state_space import CC00_Dubins_State_Space
        return CC00_Dubins_State_Space(kappa_max, sigma_max, discretization, True)

    if path_type == PathType.CC0PM_DUBINS:
        from steering_functions.cc_dubins_state_space import CC0pm_Dubins_State_Space
        return CC0pm_Dubins_State_Space(kappa_max, sigma_max, discretization, True)

    if path_type == PathType.CCPM0_DUBINS:
        from steering_functions.cc_dubins_state_space import CCpm0_Dubins_State_Space
        return CCpm0_Dubins_State_Space(kappa_max, sigma_max, discretization, True)

    if path_type == PathType.CCPMPM_DUBINS:
        from steering_functions.cc_dubins_state_space import CCpmpm_Dubins_State_Space
        return CCpmpm_Dubins_State_Space(kappa_max, sigma_max, discretization, True)

    if path_type == PathType.CC00_RS:
        from steering_functions.cc_reeds_shepp_state_space import CC00_Reeds_Shepp_State_Space
        return CC00_Reeds_Shepp_State_Space(kappa_max, sigma_max, discretization)

    if path_type == PathType.HC_RS:
        from steering_functions.hc_reeds_shepp_state_space import HC_Reeds_Shepp_State_Space
        return HC_Reeds_Shepp_State_Space(kappa_max, sigma_max, discretization)

    if path_type == PathType.HC00_RS:
        from steering_functions.hc_reeds_shepp_state_space import HC00_Reeds_Shepp_State_Space
        return HC00_Reeds_Shepp_State_Space(kappa_max, sigma_max, discretization)

    if path_type == PathType.HC0PM_RS:
        from steering_functions.hc_reeds_shepp_state_space import HC0pm_Reeds_Shepp_State_Space
        return HC0pm_Reeds_Shepp_State_Space(kappa_max, sigma_max, discretization)

    if path_type == PathType.HCPM0_RS:
        from steering_functions.hc_reeds_shepp_state_space import HCpm0_Reeds_Shepp_State_Space
        return HCpm0_Reeds_Shepp_State_Space(kappa_max, sigma_max, discretization)

    if path_type == PathType.HCPMPM_RS:
        from steering_functions.hc_reeds_shepp_state_space import HCpmpm_Reeds_Shepp_State_Space
        return HCpmpm_Reeds_Shepp_State_Space(kappa_max, sigma_max, discretization)

    # Default fallback
    from steering_functions.reeds_shepp_state_space import ReedsSheppStateSpace
    return ReedsSheppStateSpace(kappa_max, discretization)


class SteeringPath:
    """High-level wrapper for steering path computation.

    Provides a unified interface to compute optimal (shortest) or all possible
    paths between two states using the specified steering function algorithm.
    """

    def __init__(self, path_type, kappa_max, sigma_max, discretization):
        if kappa_max <= 0 or sigma_max < 0 or discretization <= 0:
            raise ValueError("Invalid parameters in SteeringPath constructor")

        self._path_type = path_type
        self._discretization = discretization
        self._kappa_max = kappa_max
        self._sigma_max = sigma_max
        self._planner = _create_planner(path_type, kappa_max, sigma_max, discretization)
        if self._planner is None:
            raise RuntimeError("Failed to create path planner")

    # --- Optimal (shortest) path interface ---

    def compute_shortest_control_sequence(self, start, goal):
        """Get the control sequence for the shortest path."""
        return self._planner.get_controls(start, goal)

    def compute_shortest_path(self, start, goal):
        """Get the discretized shortest path."""
        return self._planner.get_path(start, goal)

    # --- All paths interface ---

    def compute_all_control_sequences(self, start, goal):
        """Get all possible control sequences between two states."""
        return self._planner.get_all_controls(start, goal)

    def compute_all_paths(self, start, goal):
        """Get all possible discretized paths between two states."""
        return self._planner.get_all_paths(start, goal)

    # --- Getters ---

    @property
    def path_type(self):
        return self._path_type

    @property
    def discretization(self):
        return self._discretization

    @property
    def kappa_max(self):
        return self._kappa_max

    @property
    def sigma_max(self):
        return self._sigma_max
